esmvaltool/diag_scripts/ocean3d/ocean3d.py

```diff
@@ -0,0 +1,344 @@
+"""APPLICATE/TRR Ocean Diagnostics."""
+import itertools
+import logging
+import os
+import pickle
+from collections import OrderedDict
+
+import cartopy.crs as ccrs
+import matplotlib as mpl
+from matplotlib import cm
+import numpy as np
+
+from esmvaltool.diag_scripts.ocean3d.getdata import (aw_core, hofm_data,
+                                                     transect_data,
+                                                     tsplot_data)
+from esmvaltool.diag_scripts.ocean3d.plotting import (
+    hofm_plot, plot2d_bias, plot2d_original_grid, plot_aw_core_stat,
+    plot_profile, transect_map, transect_plot, tsplot_plot)
+from esmvaltool.diag_scripts.ocean3d.utils import (find_observations_name,
+                                                   get_clim_model_filenames,
+                                                   get_cmap, get_fx_filenames,
+                                                   timmean)
+from esmvaltool.diag_scripts.shared import run_diagnostic
+
+mpl.use('agg')  # noqa
+
+logger = logging.getLogger(os.path.basename(__file__))
+
+
+def main(cfg):
+    """Compute the time average for each input model."""
```

Comment:
This must be broken up into smaller functions. As a rule of thumb, no function should be longer than 50 lines, most even shorter.

Maybe it would be better to split this into individual diag scripts?

Reply:
As an intermediate step, I will try to break it into smaller functions. You a right that it would be better to have each of the sub diagnostics as individual scripts, but now some diagnostics use results of the others and splitting would require unnecessary heavy recalculation of many things. I have to think of a better design to make diagnostics more independent, but avoid recalculations.